### DISCUSSION TITLE
Moving COMReference to NuGet Package

### DISCUSSION THREAD

**Discussion by Cjewett** [NuGet/Home#12935] [Q&A] [answered]
Is there a recommended way to move COM References to PackageReference? We've got some custom dlls we use via COM References which requires all of our developers to register them during our machine setup process and we'd like to move them to NuGet packages. 

If you have the following:
```
<COMReference Include="KnifeSharpener">
    <Guid>{SUPER-FANCY-GUID}</Guid>
    <VersionMajor>1</VersionMajor>
    <VersionMinor>0</VersionMinor>
    <Lcid>0</Lcid>
    <WrapperTool>tlbimp<tlbimp>
    <Isolated>false</Isolated>
    <EmbedInteropTypes>true</EmbedInteropTypes>
</COMReference>
```

You can create an Interop dll by running tlbimp "KnifeSharpener.dll" /out:"KnifeSharper.Interop.dll"

I structured the NuGet package similar to this:
```
KnifeSharpener.nupkg
|-- runtimes
|   |-- win-x64
|       |-- native
|            |-- KnifeSharper.dll
|            |-- KnifeSharper.Interop.dll
|   |-- win-x86
|       |-- native
|            |-- KnifeSharper.dll
|            |-- KnifeSharper.Interop.dll
|-- build
|   |-- net48
|       |-- KnifeSharper.props
|       |-- KnifeSharper.targets
```

I'm confused how you're supposed to ensure the KnifeSharper.dll is registered or available to the application at runtime without registering it manually ahead of time though. I've looked at things like side-by-side assemblies, RegFree COM, and other resources which seem relevant but still struggling to understand how that registration piece happens. Anyone have ideas?

**Comment by zivkan** [comment, marked as answer]
NuGet doesn't have any knowledge of, or features for, COM. Therefore COM registration it outside the scope of what NuGet is designed for.

Even if it did, as a build tool, once you build your app and deploy it to a server, or ship it to a customer's machine, the COM object would still need to be registered.

**Reply by Cjewett** [reply]
@zivkan Thanks for the reply! Figured this might be the case but was hoping there was some auto-register recommendation or similar functionality. We'll most likely look into porting these DLLs so they no longer need to be registered instead.
